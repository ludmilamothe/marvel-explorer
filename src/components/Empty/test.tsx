import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import Empty from '.';

const datas = {
  imagePath: '/images/no-search.svg',
  imageAlt:
    'A magnifying glass with sad face and with some bubbles in the background',
  children: 'Some text',
};

describe('<Empty />', () => {
  it('should render the empty correct', () => {
    renderWithProviders(<Empty {...datas} />);

    const image = screen.getByRole('img', { name: datas.imageAlt });
    const children = screen.getByText(datas.children);

    expect(children).toBeInTheDocument();

    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', datas.imagePath);
  });

  it('should render the empty styles correct', () => {
    const { container } = renderWithProviders(<Empty {...datas} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
