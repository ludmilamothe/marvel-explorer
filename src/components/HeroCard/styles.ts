import styled, { css } from 'styled-components';
import media from 'styled-media-query';

export const Container = styled.div`
  ${({ theme }) => css`
    width: 12.8rem;
    height: 31.5rem;
    background-color: ${theme.colors.white};
    display: flex;
    flex-direction: column;
    overflow: hidden;
    position: relative;

    &::after {
      content: '';
      position: absolute;
      bottom: 0;
      right: 0;
      border-width: 12px;
      border-style: solid;
      border-bottom-color: ${theme.colors.black};
      border-right-color: ${theme.colors.black};
      border-top-color: transparent;
      border-left-color: transparent;
    }

    ${media.greaterThan('medium')`
      width: 100%;
      max-width: 17.7rem;
    `}
  `}
`;

export const HeroImage = styled.img`
  width: 100%;
  height: 17rem;
  flex-shrink: 0;
  object-fit: cover;
`;

export const HeroContent = styled.div`
  ${({ theme }) => css`
    width: 100%;
    height: 100%;
    position: relative;
    padding: ${theme.spacings.xsmall} ${theme.spacings.xxsmall};

    &::after {
      content: '';
      width: 100%;
      position: absolute;
      top: 0;
      right: 0;
      display: block;
      height: 4px;
      background-color: ${theme.colors.primary};
      transition: height 0.2s;
      cursor: pointer;
    }

    &:hover {
      &::after {
        height: 100%;
      }

      ${IconWrapper} {
        color: ${theme.colors.white};
        cursor: pointer;
      }

      ${HeroName} {
        color: ${theme.colors.white} !important;
        pointer-events: none;
      }
    }
  `}
`;

export const IconWrapper = styled.div`
  ${({ theme }) => css`
    width: 2.4rem;
    height: 2.4rem;
    color: ${theme.colors.primary};
    position: absolute;
    bottom: 2.4rem;
    right: 2.4rem;
    cursor: pointer;
    z-index: 10;

    & * {
      cursor: pointer;
    }
  `}
`;

export const HeroName = styled.h3`
  ${({ theme }) => css`
    position: relative;
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.black};
    text-transform: uppercase;
    z-index: 10;
  `}
`;
