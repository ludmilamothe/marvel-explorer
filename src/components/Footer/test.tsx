import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/test/helpers';

import Footer from '.';

describe('<Footer />', () => {
  it('should render the footer content correct', () => {
    renderWithTheme(<Footer />);

    const linkText = 'Mauro de Souza';
    const copy = 'Data provided by Marvel. Â© 2020 MARVEL';

    const linkElement = screen.getByRole('link', { name: linkText });

    expect(linkElement).toHaveAttribute(
      'href',
      'https://www.linkedin.com/in/maurodesouzaa/'
    );

    expect(linkElement).toHaveAttribute('target', '_blank');

    expect(screen.getByText(copy)).toBeInTheDocument();
  });

  it('should render the footer styles correct', () => {
    const { container } = renderWithTheme(<Footer />);

    const footer = container.firstChild;

    expect(footer).toMatchSnapshot();
  });
});
