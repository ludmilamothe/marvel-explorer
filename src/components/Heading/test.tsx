import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/test/helpers';

import theme from 'styles/theme/default';
import Heading from '.';

describe('<Heading />', () => {
  it('should render the heading with children', () => {
    const text = 'Just a text to test';

    renderWithTheme(<Heading>{text}</Heading>);

    const heading = screen.getByRole('heading', { name: text });

    expect(heading).toBeInTheDocument();

    expect(heading).toMatchInlineSnapshot(`
      .c0 {
        color: #FFF;
        font-weight: 900;
        font-size: 2.4rem;
        max-width: 55rem;
        text-align: center;
      }

      @media (min-width:768px) {
        .c0 {
          font-size: 4.0rem;
        }
      }

      <h1
        class="c0"
      >
        Just a text to test
      </h1>
    `);
  });

  it('should render the heading with correct styles', () => {
    const text = 'Just a text to test';

    renderWithTheme(<Heading>{text}</Heading>);

    const heading = screen.getByRole('heading', { name: text });

    expect(heading).toHaveStyle({
      fontSize: theme.font.sizes.large,
      fontWeight: theme.font.xbold,
      color: theme.colors.white,
    });

    expect(heading).toHaveStyleRule('font-size', theme.font.sizes.xlarge, {
      media: '(min-width: 768px)',
    });
  });
});
