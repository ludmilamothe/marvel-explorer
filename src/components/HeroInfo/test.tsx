import { fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import HeroInfo from '.';

const props = {
  id: 123,
  name: 'iron man',
  description:
    'Inventor Tony Stark applies his genius for high-tech solutions to problems as Iron Man, the armored Avenger.',
  wikiLink: 'https://google.com',
  comics: 123,
  series: 1234,
  stories: 12345,
};

describe('<HeroInfo />', () => {
  it('should render the HeroInfo correct', () => {
    renderWithProviders(<HeroInfo {...props} />);

    const link = screen.getByRole('link', { name: /see more/i });
    expect(screen.getByText(props.name)).toBeInTheDocument();

    expect(screen.getByText(props.description)).toBeInTheDocument();
    expect(screen.getByText(props.comics)).toBeInTheDocument();

    expect(screen.getByText(props.series)).toBeInTheDocument();
    expect(screen.getByText(props.stories)).toBeInTheDocument();

    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', props.wikiLink);
  });

  it('should alter icon between add/remove favorite', () => {
    renderWithProviders(<HeroInfo {...props} />);

    let addToFavoriteElement = screen.getByLabelText(/add hero to favorites/i);

    expect(addToFavoriteElement).toBeInTheDocument();
    fireEvent.click(addToFavoriteElement);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    expect(addToFavoriteElement).not.toBeInTheDocument();

    fireEvent.click(removeFromFavoriteElement);

    addToFavoriteElement = screen.getByLabelText(/add hero to favorites/i);

    expect(removeFromFavoriteElement).not.toBeInTheDocument();
    expect(addToFavoriteElement).toBeInTheDocument();
  });
});
