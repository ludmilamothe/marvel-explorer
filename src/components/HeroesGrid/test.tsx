import { renderWithProviders } from 'utils/test/helpers';

import HeroesGrid from '.';

describe('<HeroesGrid />', () => {
  it('should render the HeroesGrid styles correct', () => {
    const { container } = renderWithProviders(<HeroesGrid />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
