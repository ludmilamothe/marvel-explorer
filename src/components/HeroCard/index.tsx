import { useCallback, useState } from 'react';
import { useRouter } from 'next/router';

import { Favorite as FavoriteIcon } from '@styled-icons/material/Favorite';
import { FavoriteBorder as FavoteBorderIcon } from '@styled-icons/material/FavoriteBorder';

import { useFavoriteHeroes, HeroData } from 'hooks/useFavoriteHeroes';
import * as S from './styles';
import Image from 'next/image';

export type HeroCardProps = HeroData;

const HeroCard = ({ heroId, heroName, heroThumbnail }: HeroCardProps) => {
  const {
    isFavoriteHero,
    removeFavoriteHero,
    addFavoriteHero,
  } = useFavoriteHeroes();

  const { push } = useRouter();

  const [isFavorite, setIsFavorite] = useState(isFavoriteHero(heroId));

  const handleAddToFavorite = useCallback(
    (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
      e.stopPropagation();

      const heroData = {
        heroId,
        heroName,
        heroThumbnail,
      };

      addFavoriteHero(heroData);
      setIsFavorite(true);
    },
    [heroId, heroName, heroThumbnail, addFavoriteHero]
  );

  const handleRemoveFromFavorite = useCallback(
    (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
      e.stopPropagation();

      removeFavoriteHero(heroId);
      setIsFavorite(false);
    },
    [heroId, removeFavoriteHero]
  );

  return (
    <S.Container onClick={() => push(`/characters/${heroId}`)}>
      <S.ImageWrapper>
        <Image
          src={heroThumbnail}
          layout="fill"
          objectFit="cover"
          alt={heroName}
        />
      </S.ImageWrapper>
      <S.HeroContent>
        <S.HeroName>{heroName}</S.HeroName>

        {isFavorite ? (
          <S.IconWrapper onClick={handleRemoveFromFavorite}>
            <FavoteBorderIcon aria-label="Remove hero to favorites" />
          </S.IconWrapper>
        ) : (
          <S.IconWrapper onClick={handleAddToFavorite}>
            <FavoriteIcon aria-label="Add hero to favorites" />
          </S.IconWrapper>
        )}
      </S.HeroContent>
    </S.Container>
  );
};

export default HeroCard;
