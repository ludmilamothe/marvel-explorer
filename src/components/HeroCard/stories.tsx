import { Meta, Story } from '@storybook/react/types-6-0';

import HeroCard, { HeroCardProps } from '.';

export default {
  title: 'HeroCard',
  component: HeroCard,
  args: {
    heroId: 123,
    heroName: 'Iron man',
    heroThumbnail:
      'http://i.annihil.us/u/prod/marvel/i/mg/9/c0/527bb7b37ff55.jpg',
  },
} as Meta<HeroCardProps>;

export const Basic: Story<HeroCardProps> = args => <HeroCard {...args} />;
