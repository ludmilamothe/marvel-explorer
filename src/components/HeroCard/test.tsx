/* eslint-disable @typescript-eslint/no-var-requires */

import { fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { storageKey } from 'hooks/useFavoriteHeroes';
import HeroCard from '.';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/router'), 'useRouter');
const push = jest.fn();

useRouter.mockImplementation(() => ({
  push,
}));

const heroData = {
  heroId: 123,
  heroName: 'Iron man',
  heroThumbnail:
    'http://i.annihil.us/u/prod/marvel/i/mg/9/c0/527bb7b37ff55.jpg',
};

describe('<HeroCard />', () => {
  beforeEach(() => {
    window.localStorage.clear();
  });

  it('should render the hero card correct', () => {
    renderWithProviders(<HeroCard {...heroData} />);

    expect(
      screen.getByRole('img', { name: heroData.heroName })
    ).toBeInTheDocument();

    expect(
      screen.getByRole('heading', { name: heroData.heroName })
    ).toBeInTheDocument();

    expect(screen.getByLabelText(/add hero to favorites/i)).toBeInTheDocument();
  });

  it('should render the hero card styles correct', () => {
    const { container } = renderWithProviders(<HeroCard {...heroData} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should add hero to favorites', () => {
    renderWithProviders(<HeroCard {...heroData} />);

    const addToFavoriteElement = screen.getByLabelText(
      /add hero to favorites/i
    );

    expect(addToFavoriteElement).toBeInTheDocument();
    fireEvent.click(addToFavoriteElement);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    expect(addToFavoriteElement).not.toBeInTheDocument();

    const storagedHero = JSON.parse(window.localStorage.getItem(storageKey)!);

    expect(storagedHero).toEqual([heroData]);
  });

  it('should remove hero to favorites', () => {
    window.localStorage.setItem(storageKey, JSON.stringify([heroData]));

    renderWithProviders(<HeroCard {...heroData} />);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    fireEvent.click(removeFromFavoriteElement);

    const addToFavoriteElement = screen.getByLabelText(
      /add hero to favorites/i
    );

    expect(removeFromFavoriteElement).not.toBeInTheDocument();
    expect(addToFavoriteElement).toBeInTheDocument();

    const storagedHero = JSON.parse(window.localStorage.getItem(storageKey)!);
    expect(storagedHero).toEqual([]);
  });

  it('should navigate to hero page', () => {
    const { container } = renderWithProviders(<HeroCard {...heroData} />);

    const heroCard = container.firstChild!;
    fireEvent.click(heroCard);

    expect(push).toBeCalledTimes(1);
    expect(push).toBeCalledWith(`/characters/${heroData.heroId}`);
  });
});
