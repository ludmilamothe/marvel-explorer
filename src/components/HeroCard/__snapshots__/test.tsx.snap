// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeroCard /> should render the hero card styles correct 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  width: 100%;
  height: 31.5rem;
  background-color: #FFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
  position: relative;
}

.c0::after {
  content: '';
  position: absolute;
  bottom: 0;
  right: 0;
  border-width: 12px;
  border-style: solid;
  border-bottom-color: #202020;
  border-right-color: #202020;
  border-top-color: transparent;
  border-left-color: transparent;
}

.c1 {
  width: 100%;
  height: 17rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  object-fit: cover;
}

.c2 {
  width: 100%;
  height: 100%;
  position: relative;
  padding: 1.2rem 0.8rem;
}

.c2::after {
  content: '';
  width: 100%;
  position: absolute;
  top: 0;
  right: 0;
  display: block;
  height: 4px;
  background-color: #E62429;
  -webkit-transition: height 0.2s;
  transition: height 0.2s;
  cursor: pointer;
}

.c2:hover::after {
  height: 100%;
}

.c2:hover .c5 {
  color: #FFF;
  cursor: pointer;
}

.c2:hover .c3 {
  color: #FFF !important;
  pointer-events: none;
}

.c6 {
  width: 2.4rem;
  height: 2.4rem;
  color: #E62429;
  position: absolute;
  bottom: 2.4rem;
  right: 2.4rem;
  cursor: pointer;
  z-index: 10;
}

.c6 * {
  cursor: pointer;
}

.c4 {
  position: relative;
  font-size: 1.2rem;
  font-weight: 700;
  color: #202020;
  text-transform: uppercase;
  z-index: 10;
}

<div
  class="c0"
>
  <img
    alt="Iron man"
    class="c1"
    src="http://i.annihil.us/u/prod/marvel/i/mg/9/c0/527bb7b37ff55.jpg"
  />
  <div
    class="c2"
  >
    <h3
      class="c3 c4"
    >
      Iron man
    </h3>
    <div
      class="c5 c6"
    >
      <svg
        aria-hidden="true"
        aria-label="Add hero to favorites"
        class="c7"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
        />
      </svg>
    </div>
  </div>
</div>
`;
